### PR TITLE
Dala 2978 fix document uploads in upload forms

### PR DESCRIPTION
**Pull request by fsoffel**
# Pull Request \<Title>
`<Description here>`
## Things to do during Peer Review
Please check all boxes before the Pull Request is merged. In case you skip a box, describe in the PRs description (that means: here) why the check is skipped.
- [x] The Github Actions (including Sonarqube Gateway and Lint Checks) are green. This is enforced by Github. 
- [x] A peer-review has been executed
  - [x] The code has been manually inspected by someone who did not implement the feature
- [x] The PR actually implements what is described in the JIRA-Issue
- [x] At least one test exists testing the new feature
  - [x] If you have created new test files, make sure that they are included in a test container and actually run in the CI
- [x] Documentation is updated as required
- [x] The automated deployment is updated if required
- [ ] If there was a database entity class added, there must also be a migration script for creating the corresponding database if flyway is already used by the service
- [ ] IF there are changes done to the framework data models or to a database entity class, the following steps were completed in order
  - [ ] The version of main currently active on prod is deployed to a dev server with `Reset non-user related Docker Volumes & Re-populate` turned on
  - [ ] It's verified that the CD run is green
  - [ ] The data from prod is migrated using `./migrateData.sh dataland.com <SOURCE_API_KEY> <TARGET> <TARGET_API_KEY>` 
  - [ ] The feature branch is deployed to the same server with `Reset non-user related Docker Volumes & Re-populate` turned off
  - [ ] It's verified that the CD run is green
  - [ ] The new feature is manually used/tested/observed on the dev server
- [x] ELSE, the new version is deployed to the dev server "dev1" using this branch
  - [x] Run with setting `Reset non-user related Docker Volumes & Re-populate` turned on 
  - [x] It's verified that this version actually is the one deployed (check gitinfo for branch name and commit id!)
  - [x] It's verified that the CD run is green
  - [x] The new feature is manually used/tested/observed on dev server
- [x] Make sure that pull requests in the development tools repository, which are connected to this pull requests, 
are merged

### CODE REVIEW

File: dataland-frontend/src/components/forms/CreateLksgDataset.vue [hunk 1]
```diff
@@ -52,7 +52,9 @@
                           :validation-label="field.validationLabel"
                           :data-test="field.name"
                           :shouldDisableCheckboxes="true"
-                          @reportsUpdated="updateDocumentList"
+                          @field-specific-documents-updated="
+                            updateDocumentList(`${category.name}.${subcategory.name}.${field.name}`, $event)
```
Comment (stewal19) [inline dataland-frontend/src/components/forms/CreateLksgDataset.vue:56]: Why this addition?

File: dataland-frontend/src/components/forms/CreateP2pDataset.vue [hunk 2]
```diff
@@ -243,6 +267,10 @@ export default defineComponent({
     async postP2pData(): Promise<void> {
       this.messageCounter++;
       try {
+        this.filterFieldSpecificDocumentsByEnabledFieldOwnership();
```
Comment (stewal19) [inline dataland-frontend/src/components/forms/CreateP2pDataset.vue:270]: Why do we need this?

File: dataland-frontend/src/components/forms/CreateP2pDataset.vue [hunk 3]
```diff
@@ -264,6 +292,34 @@ export default defineComponent({
         this.postP2pDataProcessed = true;
       }
     },
+
```
Comment (stewal19) [inline dataland-frontend/src/components/forms/CreateP2pDataset.vue:295]: Can someone explain these changes to me?

File: dataland-frontend/src/components/forms/parts/UploadReports.vue [hunk 4]
```diff
@@ -1,71 +1,73 @@
...
-  <!-- Select company reports -->
-  <div :class="isEuTaxonomy ? 'col-9 formFields' : 'formField'">
-    <h3 class="mt-0">Select company reports</h3>
-    <UploadDocumentsForm
-      ref="uploadDocumentsForm"
-      @updatedDocumentsSelectedForUpload="handleUpdatedDocumentsSelectedForUpload"
-      :name="name"
-    />
-  </div>
+  <div :data-test="`upload-reports-${name}`">
```
Comment (stewal19) [inline dataland-frontend/src/components/forms/parts/UploadReports.vue:2]: Here just this test-marker was introduced, correct?

File: dataland-frontend/tests/component/components/forms/parts/elements/basic/UploadDocumentsForm.cy.ts [hunk 5]
```diff
@@ -1,8 +1,10 @@
 import UploadDocumentsForm from "@/components/forms/parts/elements/basic/UploadDocumentsForm.vue";
-import { uploadDocuments } from "@sharedUtils/components/UploadDocuments";
```
Comment (stewal19) [inline dataland-frontend/tests/component/components/forms/parts/elements/basic/UploadDocumentsForm.cy.ts:2]: Why is it now UploadDocuments and no longer uploadDocuments?

File: dataland-frontend/tests/sharedUtils/components/UploadDocuments.ts [hunk 6]
```diff
@@ -1,119 +1,58 @@
 import { TEST_PDF_FILE_BASEPATH } from "@sharedUtils/ConstantsForPdfs";
 
-export const uploadDocuments = {
```
[the remark this answers is not in the document]
Comment (stewal19) [reply dataland-frontend/tests/sharedUtils/components/UploadDocuments.ts:3]: Or are those just moved to UploadReports?

File: dataland-frontend/tests/e2e/utils/SfdrUpload.ts [hunk 7]
```diff
@@ -1,11 +1,12 @@
-import { uploadDocuments } from "@sharedUtils/components/UploadDocuments";
 import { TEST_PDF_FILE_NAME } from "@sharedUtils/ConstantsForPdfs";
+import { UploadReports } from "@sharedUtils/components/UploadReports";
 
 /**
  * Adds reports to the dataset via the Sfdr upload form for the given dataset
  */
 export function selectsReportsForUploadInSfdrForm(): void {
-  uploadDocuments.selectFile(TEST_PDF_FILE_NAME, "referencedReports");
```
Comment (stewal19) [inline dataland-frontend/tests/e2e/utils/SfdrUpload.ts:8]: This looks logic-wise very similar to EuTaxonomyNonFinancialsUpload. Would it have merit to further generalize it to reduce code duplications?

File: dataland-frontend/src/utils/FileUploadUtils.ts [hunk 8]
```diff
@@ -25,10 +25,15 @@ export async function uploadFiles(
   getKeycloakPromise: () => Promise<Keycloak>,
 ): Promise<void> {
   const documentControllerApi = new ApiClientProvider(getKeycloakPromise()).apiClients.documentController;
+  const alreadyUploadedFileReferences = new Set<string>();
   for (const fileToUpload of files) {
+    if (alreadyUploadedFileReferences.has(fileToUpload.fileReference)) {
```
Comment (stewal19) [inline dataland-frontend/src/utils/FileUploadUtils.ts:30]: Why is this added?

Comment (stewal19) [reply dataland-frontend/src/utils/FileUploadUtils.ts:30]: Maybe we can talk about this in a short call so i get a better understanding what was changed here.

File: dataland-frontend/tests/component/components/forms/UploadFieldSpecificReports.cy.ts [hunk 9]
```diff
@@ -0,0 +1,176 @@
+import CreateSfdrDataset from "@/components/forms/CreateSfdrDataset.vue";
```
Comment (stewal19) [inline dataland-frontend/tests/component/components/forms/UploadFieldSpecificReports.cy.ts:1]: Do we need one test for SRDR and one for P2P?

File: dataland-frontend/tests/component/components/forms/UploadFieldSpecificReports.cy.ts [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+    createP2pDataset.selectSector("Livestock Farming");
+    cy.wait(100);
+    submitButton.buttonAppearsEnabled();
+    submitButton.clickButton();
+
+    setOfHashesThatShouldBeCheckedForExistence.forEach((hash) => {
+      cy.wait(`@documentExists-${hash}`);
+    });
+    cy.wait(100);
+    cy.get("@documentExists").should("have.been.calledOnce");
```
Comment (stewal19) [inline dataland-frontend/tests/component/components/forms/UploadFieldSpecificReports.cy.ts:174]: Same as above

File: dataland-frontend/src/components/forms/CreateSfdrDataset.vue [hunk 11]
```diff
@@ -309,9 +317,22 @@ export default defineComponent({
      */
     updateDocumentsList(reportsNamesAndReferences: object, reportsToUpload: DocumentToUpload[]) {
       this.namesAndReferencesOfAllCompanyReportsForTheDataset = reportsNamesAndReferences;
-      this.documents = new Map();
-      if (reportsToUpload?.length) {
-        reportsToUpload.forEach((document) => this.documents.set(document.file.name, document));
+      if (reportsToUpload.length) {
```
Comment (stewal19) [inline dataland-frontend/src/components/forms/CreateSfdrDataset.vue:320]: Why do we need extra logic for the create pages?

File: dataland-frontend/src/components/forms/CreateLksgDataset.vue (hunk 1 above)
Comment (fsoffel) [reply dataland-frontend/src/components/forms/CreateLksgDataset.vue:56]: I changed the name in the base datapoint to something more fitting. And in lksg this is only an event in base data points. Thus I renamed it.

File: dataland-frontend/src/components/forms/CreateP2pDataset.vue (hunk 2 above)
Comment (fsoffel) [reply dataland-frontend/src/components/forms/CreateP2pDataset.vue:270]: P2P has these sectors which can be enabled and disabled. If I e.g. enable LivestockFarming and put a document into a base datapoint in the LivestockFarming sector that document should be uploaded. But if I disable the sector again it should no longer be uploaded because the corresponding base data point does no longer exist. These fields are filtered out in this function.

File: dataland-frontend/src/components/forms/CreateP2pDataset.vue (hunk 3 above)
Comment (fsoffel) [reply dataland-frontend/src/components/forms/CreateP2pDataset.vue:295]: They are necessary to have this page also upload documents. You can call me for further explanation.

File: dataland-frontend/src/components/forms/parts/UploadReports.vue (hunk 4 above)
Comment (fsoffel) [reply dataland-frontend/src/components/forms/parts/UploadReports.vue:2]: Yes.

File: dataland-frontend/tests/component/components/forms/parts/elements/basic/UploadDocumentsForm.cy.ts (hunk 5 above)
Comment (fsoffel) [reply dataland-frontend/tests/component/components/forms/parts/elements/basic/UploadDocumentsForm.cy.ts:2]: I made it a class instead of an object because due to multiple base data points in a form a single object to manage access to all UploadDocumentForms made not much sense to me. Now it's one object per UploadDocumentsForm.

File: dataland-frontend/tests/sharedUtils/components/UploadDocuments.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (fsoffel) [reply dataland-frontend/tests/sharedUtils/components/UploadDocuments.ts:3]: They got moved to UploadReports.ts because they are specific to the UploadReports component. And UploadDocuments here is supposed to only mirror the UploadDocumentsForm component.

File: dataland-frontend/src/utils/FileUploadUtils.ts (hunk 8 above)
Comment (fsoffel) [reply dataland-frontend/src/utils/FileUploadUtils.ts:30]: Of cours. However a short explanation: In a Create*DataModel.vue file we filtered for documents with the same has to not upload them twice and each of these used this function to upload the documents. The idea was before doing this filtering for each creation form we can do it at this centralized place such that noone must ever bother with it again.

File: dataland-frontend/tests/component/components/forms/UploadFieldSpecificReports.cy.ts (hunk 9 above)
Comment (fsoffel) [reply dataland-frontend/tests/component/components/forms/UploadFieldSpecificReports.cy.ts:1]: Yes. Because we had two different bugs in both and I think this is one of our rules. Bug => Test for bug such that it never occurs again.

File: dataland-frontend/tests/component/components/forms/UploadFieldSpecificReports.cy.ts [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+    cy.get('div[data-test="environmentalPolicy"] button .pi-times').click();
+    cy.wait(100);
+    submitButton.buttonAppearsEnabled();
+    submitButton.clickButton();
+
+    setOfHashesThatShouldBeCheckedForExistence.forEach((hash) => {
+      cy.wait(`@documentExists-${hash}`);
+    });
+    cy.wait(100);
+    cy.get("@documentExists").should("have.been.calledOnce");
```
[the remark this answers is not in the document]
Comment (fsoffel) [reply dataland-frontend/tests/component/components/forms/UploadFieldSpecificReports.cy.ts:151]: Because one "first" gets replaced by "second" and "third" gets removed entirely. The reason for this test is to make sure that file deletion works properly.

File: dataland-frontend/tests/e2e/utils/SfdrUpload.ts (hunk 7 above)
Comment (fsoffel) [reply dataland-frontend/tests/e2e/utils/SfdrUpload.ts:8]: I see your point. I feel like the tradeoff is readability here. But I have no strong opinion on this. Let me change it.

Comment (fsoffel) [reply dataland-frontend/tests/e2e/utils/SfdrUpload.ts:8]: Mislooked. This is already an extracted function and thus can already be reused as is. Thus no arguments against it at all.

File: dataland-frontend/src/components/forms/CreateP2pDataset.vue [hunk 13]
```diff
@@ -53,6 +53,9 @@
                           :validation-label="field.validationLabel"
                           :data-test="field.name"
                           :ref="field.name"
+                          @field-specific-documents-updated="
```
[the remark this answers is not in the document]
Comment (stewal19) [reply dataland-frontend/src/components/forms/CreateP2pDataset.vue:56]: For real? How have we missed that?

File: dataland-frontend/tests/sharedUtils/components/UploadDocuments.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (stewal19) [reply dataland-frontend/tests/sharedUtils/components/UploadDocuments.ts:3]: ok

File: dataland-frontend/tests/component/components/forms/UploadFieldSpecificReports.cy.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (stewal19) [reply dataland-frontend/tests/component/components/forms/UploadFieldSpecificReports.cy.ts:151]: ok

File: dataland-frontend/tests/component/components/forms/UploadFieldSpecificReports.cy.ts (hunk 10 above)
Comment (stewal19) [reply dataland-frontend/tests/component/components/forms/UploadFieldSpecificReports.cy.ts:174]: ok

File: dataland-frontend/tests/component/components/forms/parts/elements/basic/UploadDocumentsForm.cy.ts (hunk 5 above)
Comment (stewal19) [reply dataland-frontend/tests/component/components/forms/parts/elements/basic/UploadDocumentsForm.cy.ts:2]: ok

File: dataland-frontend/src/utils/FileUploadUtils.ts (hunk 8 above)
Comment (stewal19) [reply dataland-frontend/src/utils/FileUploadUtils.ts:30]: ok, thank you for your clarification

File: dataland-frontend/src/components/forms/CreateSfdrDataset.vue (hunk 11 above)
Comment (stewal19) [reply dataland-frontend/src/components/forms/CreateSfdrDataset.vue:320]: okk

File: dataland-frontend/src/components/forms/CreateLksgDataset.vue (hunk 1 above)
Comment (stewal19) [reply dataland-frontend/src/components/forms/CreateLksgDataset.vue:56]: ok

File: dataland-frontend/src/components/forms/CreateP2pDataset.vue (hunk 2 above)
Comment (stewal19) [reply dataland-frontend/src/components/forms/CreateP2pDataset.vue:270]: ok

File: dataland-frontend/src/components/forms/CreateP2pDataset.vue (hunk 3 above)
Comment (stewal19) [reply dataland-frontend/src/components/forms/CreateP2pDataset.vue:276]: Bugs were not encountered on the deployed version. So this seems to work and iam closing this here
